### PR TITLE
FIREWALL: Use rich rules instead of direct rules

### PR DESCRIPTION
**Pull request by aplopez**
Direct rules have been deprecated for some time now, so we switch to using rich rules.

However, we need a specific policy to which the rule will be applied. This policy is permanent (a requirement) and is created at setup, deleted at teardown, and remains active during the whole test. Only the rules are dynamic.

This works on RHEL 8 thourgh 10.

### CODE REVIEW

File: pytest_mh/utils/firewall.py [hunk 1]
```diff
@@ -244,12 +245,17 @@ def __init__(self, host: MultihostHost) -> None:
         not remove "accept" rule but the "drop" rule takes precedence.
         """
 
+        self._policy: str = f"test-policy-{randrange(99999)}"
+        self.add_policy(self._policy, ingress="HOST", egress="ANY")
+        self.host.ssh.exec(["firewall-cmd", "--reload"], log_level=SSHLog.Error)
```
Comment (pbrezina) [inline pytest_mh/utils/firewall.py:250]: Please, move this to setup().

Comment (aplopez) [reply pytest_mh/utils/firewall.py:257]: I should have guessed that there was a `setup()` function if there was a `teardown()` one.
Done.

File: pytest_mh/utils/firewall.py [hunk 2]
```diff
@@ -289,95 +302,77 @@ def _next_priority(self) -> int:
...
-        :type table: str, optional
-        :param ip_family: If the rules is added as IPv4, IPv6 rule or both, defaults to ``all``.
-        :type ip_family: Literal["ipv4", "ipv6", "all"], optional
-        :param priority: Rule priority, defaults to None (= auto-assign next value)
-        :type priority: int | None, optional
-        :return: Rule priority, to be used for rule removal.
+        Add a new (permanent) policy.
+
+        Except for the name, all parameters are optional. When the priority is not provided,
+        the next priority is assigned. When the other parameters is not provided,
```
Comment (ikerexxe) [inline pytest_mh/utils/firewall.py:318]: s/is/are

File: pytest_mh/utils/firewall.py [hunk 3]
```diff
@@ -289,95 +302,77 @@ def _next_priority(self) -> int:
         self._priority -= 1
         return priority
 
-    def add_direct_rule(
+    def add_policy(
         self,
-        chain: str,
-        args: list[Any],
+        name: str,
```
Comment (ikerexxe) [inline pytest_mh/utils/firewall.py:307]: This parameter is missing in the description

File: pytest_mh/utils/firewall.py [hunk 4]
```diff
@@ -289,95 +302,77 @@ def _next_priority(self) -> int:
...
         :rtype: int
         """
+        self.logger.info(f'Firewalld: adding policy "{name}"')
+
+        cmd = ["firewall-cmd", "--permanent", "--new-policy", name]
+        self.host.ssh.exec(cmd, log_level=SSHLog.Error)
+        cmd[2] = "--policy"
+
         if priority is None:
             priority = self._next_priority
```
Comment (ikerexxe) [inline pytest_mh/utils/firewall.py:340]: I wonder, shouldn't this variable change at some point?

File: pytest_mh/utils/firewall.py [hunk 5]
```diff
@@ -289,95 +302,77 @@ def _next_priority(self) -> int:
...
-        priority: int,
-        chain: str,
-        args: list[Any],
-        *,
-        table: str = "filter",
-        ip_family: Literal["ipv4", "ipv6", "all"],
-    ) -> None:
-        """
-        Remove direct rule.
+    def remove_policy(self, name: str) -> None:
```
Comment (ikerexxe) [inline pytest_mh/utils/firewall.py:354]: I know it's a simple method but I'd expect some comment explaining what it does and the parameters.

File: pytest_mh/utils/firewall.py [hunk 6]
```diff
@@ -541,15 +547,13 @@ def __add_port(
             else:
                 port, protocol = self.firewall.parse_port_spec(item)
 
-            self.firewall.add_direct_rule(
-                chain="OUTPUT", args=["-p", protocol, "--dport", port, "-j", action], table="filter"
-            )
+            self.firewall.add_rich_rule(f"port port={port} protocol={protocol} {action}")
```
Comment (ikerexxe) [inline pytest_mh/utils/firewall.py:550]: You wrote `port` two times. Is that correct?

**Review by ikerexxe — COMMENTED** [review]
Some inline comments, apart from that it looks good.

File: pytest_mh/utils/firewall.py (hunk 2 above)
Comment (aplopez) [reply pytest_mh/utils/firewall.py:318]: Fixed

File: pytest_mh/utils/firewall.py (hunk 3 above)
Comment (aplopez) [reply pytest_mh/utils/firewall.py:307]: Fixed

File: pytest_mh/utils/firewall.py (hunk 4 above)
Comment (aplopez) [reply pytest_mh/utils/firewall.py:340]: It's not a variable but a property. It changes every time it is used. Check line 290.

File: pytest_mh/utils/firewall.py (hunk 5 above)
Comment (aplopez) [reply pytest_mh/utils/firewall.py:354]: Done

File: pytest_mh/utils/firewall.py (hunk 4 above)
Comment (ikerexxe) [reply pytest_mh/utils/firewall.py:342]: Now I see the `@property` attribute. Thanks!

**Review by ikerexxe — APPROVED** [review]
LGTM. Thanks for all the explanations!

File: pytest_mh/utils/firewall.py [hunk 7]
```diff
@@ -289,95 +302,85 @@ def _next_priority(self) -> int:
...
-        table: str = "filter",
-        ip_family: Literal["ipv4", "ipv6", "all"] = "all",
+        ingress: str | None = None,
+        egress: str | None = None,
         priority: int | None = None,
+        target: Literal["CONTINUE", "ACCEPT", "DROP", "REJECT"] | None = None,
     ) -> int:
         """
-        Add a new direct rule.
+        Add a new (permanent) policy.
```
Comment (pbrezina) [inline pytest_mh/utils/firewall.py:315]: But it is still removed in teardown, right?

Comment (aplopez) [reply pytest_mh/utils/firewall.py:328]: I see your point. It's done now.

**Review by pbrezina — COMMENTED** [review]
I didn't review the code, but it looks ok conceptually.

As a rule of thumb - don't call any commands in __init__, use setup for that. This way, it is possible to call setup only when something from the utils is actually used (see mh_postpone_setup). And every change done must be reverted in teardown so we do not polute other tests.

**Comment by pbrezina** [comment]
Bump @thalman

**Review by thalman — APPROVED** [review]
Sorry, I did the review earlier and forgot to ACK it. Approving now...

**Comment by pbrezina** [comment]
@aplopez Please, resolve conflict and we can merge.

**Pull request merged**